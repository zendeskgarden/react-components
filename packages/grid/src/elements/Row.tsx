/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { StyledRow } from '../styled';

export interface IRowProps extends HTMLAttributes<HTMLDivElement> {
  /** Remove grid gutter spacing */
  isCollapsed?: boolean;
  /** Use flexbox alignment utilities to vertically align content */
  alignItems?: 'start' | 'center' | 'end';
  /** Use flexbox justify utilities to justify content */
  justifyContent?: 'start' | 'center' | 'end' | 'around' | 'between';
}

/**
 * Accepts all `<div>` attributes and events
 */
export const Row = React.forwardRef<HTMLDivElement, IRowProps>((props, ref) => (
  <StyledRow ref={ref} {...props} />
));

Row.propTypes = {
  isCollapsed: PropTypes.bool,
  alignItems: PropTypes.oneOf(['start', 'center', 'end']),
  justifyContent: PropTypes.oneOf(['start', 'center', 'end', 'around', 'between'])
};
