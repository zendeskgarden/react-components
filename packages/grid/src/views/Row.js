/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import PropTypes from 'prop-types';
import styled from 'styled-components';
import classNames from 'classnames';
import { retrieveTheme } from '@zendeskgarden/react-theming';
import GridStyles from '@zendeskgarden/css-grid';

const COMPONENT_ID = 'grid.row';

/**
 * Accepts all `<div>` props
 */
const Row = styled.div.attrs({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  className: props =>
    classNames(GridStyles.row, {
      [GridStyles['no-gutters']]: !props.gutters,
      [GridStyles[`align-items-${props.alignItems}`]]: props.alignItems,
      [GridStyles[`justify-content-${props.justifyContent}`]]: props.justifyContent
    })
})`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

Row.propTypes = {
  /** Enables margin for rows and padding for columns  */
  gutters: PropTypes.bool,
  /** Use flexbox alignment utilities to vertically align content */
  alignItems: PropTypes.oneOf(['start', 'center', 'end']),
  /** Use flexbox justify utilities to justify content */
  justifyContent: PropTypes.oneOf(['start', 'center', 'end', 'around', 'between'])
};

Row.defaultProps = {
  gutters: true
};

/** @component */
export default Row;
