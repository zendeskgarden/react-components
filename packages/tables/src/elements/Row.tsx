/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { useState, HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { composeEventHandlers } from '@zendeskgarden/container-utilities';
import { StyledRow, IStyledRowProps } from '../styled';
import { useTableContext } from '../utils/useTableContext';

/**
 * @extends HTMLAttributes<HTMLTableRowElement>
 */
export const Row = React.forwardRef<
  HTMLTableRowElement,
  Omit<IStyledRowProps, 'size'> & HTMLAttributes<HTMLTableRowElement>
>(({ onFocus, onBlur, isFocused: focused, ...otherProps }, ref) => {
  const [isFocused, setIsFocused] = useState(false);
  const { size, isReadOnly } = useTableContext();

  return (
    <StyledRow
      onFocus={composeEventHandlers(onFocus, () => {
        setIsFocused(true);
      })}
      onBlur={composeEventHandlers(onBlur, () => {
        setIsFocused(false);
      })}
      size={size}
      isReadOnly={isReadOnly}
      isFocused={typeof focused === 'undefined' ? isFocused : focused}
      ref={ref}
      {...otherProps}
    />
  );
});

Row.displayName = 'Row';

Row.propTypes = {
  isStriped: PropTypes.bool,
  isFocused: PropTypes.bool,
  isHovered: PropTypes.bool,
  isSelected: PropTypes.bool
};
