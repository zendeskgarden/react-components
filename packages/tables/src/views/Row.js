/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import TableStyles from '@zendeskgarden/css-tables';
import { retrieveTheme } from '@zendeskgarden/react-theming';
import { composeEventHandlers } from '@zendeskgarden/react-selection';

const COMPONENT_ID = 'tables.row';

export const StyledRow = styled.div.attrs(props => ({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  tabIndex: -1,
  role: 'row',
  className: classNames(TableStyles['c-table__row'], {
    [TableStyles['c-table__row--stripe']]: props.striped,

    [TableStyles['is-focused']]: props.focused,
    [TableStyles['is-hovered']]: props.hovered,
    [TableStyles['is-selected']]: props.selected
  })
}))`
  /* stylelint-disable */
  display: flex !important;
  height: auto !important;
  /* stylelint-enable */

  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

/**
 * Accepts all `<tr>` props
 */
// eslint-disable-next-line react/prop-types
const Row = React.forwardRef(({ onFocus, onBlur, focused, ...otherProps }, ref) => {
  const [isFocused, setIsFocused] = useState(false);

  return (
    <StyledRow
      onFocus={composeEventHandlers(onFocus, () => {
        setIsFocused(true);
      })}
      onBlur={composeEventHandlers(onBlur, () => {
        setIsFocused(false);
      })}
      focused={typeof focused === 'undefined' ? isFocused : focused}
      ref={ref}
      {...otherProps}
    />
  );
});

Row.propTypes = {
  /** Applies striped styling */
  striped: PropTypes.bool,
  /** Applies focused styling */
  focused: PropTypes.bool,
  /** Applies hovered styling */
  hovered: PropTypes.bool,
  /** Applies selected styling */
  selected: PropTypes.bool
};

export default Row;
