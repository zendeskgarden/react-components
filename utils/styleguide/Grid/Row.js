import React from 'react';
import PropTypes from 'prop-types';

import styles from './styles.css';

const Row = ({ children }) => <div className={styles.row}>{children}</div>;

Row.propTypes = {
  children: PropTypes.node.isRequired
};

export default Row;
