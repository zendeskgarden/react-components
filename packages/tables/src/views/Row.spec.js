/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { mount } from 'enzyme';

import Row from './Row';

// eslint-disable-next-line
const ExampleWrapper = ({ children }) => (
  <table>
    <tbody>{children}</tbody>
  </table>
);

describe('Row', () => {
  it('applies default styling by default', () => {
    const wrapper = mount(
      <ExampleWrapper>
        <Row />
      </ExampleWrapper>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('applies hovered styling if provided', () => {
    const wrapper = mount(
      <ExampleWrapper>
        <Row hovered />
      </ExampleWrapper>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('applies selected styling if provided', () => {
    const wrapper = mount(
      <ExampleWrapper>
        <Row selected />
      </ExampleWrapper>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('applies focused styling if provided', () => {
    const wrapper = mount(
      <ExampleWrapper>
        <Row focused />
      </ExampleWrapper>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('applies group styling if provided', () => {
    const wrapper = mount(
      <ExampleWrapper>
        <Row group />
      </ExampleWrapper>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('applies header styling if provided', () => {
    const wrapper = mount(
      <ExampleWrapper>
        <Row header />
      </ExampleWrapper>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('applies striped styling if provided', () => {
    const wrapper = mount(
      <ExampleWrapper>
        <Row striped />
      </ExampleWrapper>
    );

    expect(wrapper).toMatchSnapshot();
  });

  describe('onFocus', () => {
    it('applies focused state', () => {
      const wrapper = mount(
        <ExampleWrapper>
          <Row />
        </ExampleWrapper>
      );

      wrapper.find('tr').simulate('focus');
      expect(wrapper.find('tr').parent()).toHaveProp('focused', true);
    });
  });

  describe('onBlur', () => {
    it('removes focused state', () => {
      const wrapper = mount(
        <ExampleWrapper>
          <Row />
        </ExampleWrapper>
      );

      wrapper.find('tr').simulate('focus');
      wrapper.find('tr').simulate('blur');
      expect(wrapper.find('tr').parent()).toHaveProp('focused', false);
    });
  });
});
