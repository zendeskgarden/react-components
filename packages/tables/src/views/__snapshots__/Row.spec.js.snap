// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row applies default styling by default 1`] = `
<ExampleWrapper>
  <table>
    <tbody>
      <Row>
        <Row__StyledRow
          focused={false}
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <tr
            className="c-table__row "
            data-garden-id="tables.row"
            data-garden-version="version"
            onBlur={[Function]}
            onFocus={[Function]}
            tabIndex={-1}
          />
        </Row__StyledRow>
      </Row>
    </tbody>
  </table>
</ExampleWrapper>
`;

exports[`Row applies focused styling if provided 1`] = `
<ExampleWrapper>
  <table>
    <tbody>
      <Row
        focused={true}
      >
        <Row__StyledRow
          focused={true}
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <tr
            className="c-table__row is-focused "
            data-garden-id="tables.row"
            data-garden-version="version"
            onBlur={[Function]}
            onFocus={[Function]}
            tabIndex={-1}
          />
        </Row__StyledRow>
      </Row>
    </tbody>
  </table>
</ExampleWrapper>
`;

exports[`Row applies group styling if provided 1`] = `
<ExampleWrapper>
  <table>
    <tbody>
      <Row
        group={true}
      >
        <Row__StyledRow
          focused={false}
          group={true}
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <tr
            className="c-table__row c-table__row--group "
            data-garden-id="tables.row"
            data-garden-version="version"
            onBlur={[Function]}
            onFocus={[Function]}
            tabIndex={-1}
          />
        </Row__StyledRow>
      </Row>
    </tbody>
  </table>
</ExampleWrapper>
`;

exports[`Row applies header styling if provided 1`] = `
<ExampleWrapper>
  <table>
    <tbody>
      <Row
        header={true}
      >
        <Row__StyledRow
          focused={false}
          header={true}
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <tr
            className="c-table__row c-table__row--header "
            data-garden-id="tables.row"
            data-garden-version="version"
            onBlur={[Function]}
            onFocus={[Function]}
            tabIndex={-1}
          />
        </Row__StyledRow>
      </Row>
    </tbody>
  </table>
</ExampleWrapper>
`;

exports[`Row applies hovered styling if provided 1`] = `
<ExampleWrapper>
  <table>
    <tbody>
      <Row
        hovered={true}
      >
        <Row__StyledRow
          focused={false}
          hovered={true}
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <tr
            className="c-table__row is-hovered "
            data-garden-id="tables.row"
            data-garden-version="version"
            onBlur={[Function]}
            onFocus={[Function]}
            tabIndex={-1}
          />
        </Row__StyledRow>
      </Row>
    </tbody>
  </table>
</ExampleWrapper>
`;

exports[`Row applies selected styling if provided 1`] = `
<ExampleWrapper>
  <table>
    <tbody>
      <Row
        selected={true}
      >
        <Row__StyledRow
          focused={false}
          onBlur={[Function]}
          onFocus={[Function]}
          selected={true}
        >
          <tr
            className="c-table__row is-selected "
            data-garden-id="tables.row"
            data-garden-version="version"
            onBlur={[Function]}
            onFocus={[Function]}
            selected={true}
            tabIndex={-1}
          />
        </Row__StyledRow>
      </Row>
    </tbody>
  </table>
</ExampleWrapper>
`;

exports[`Row applies striped styling if provided 1`] = `
<ExampleWrapper>
  <table>
    <tbody>
      <Row
        striped={true}
      >
        <Row__StyledRow
          focused={false}
          onBlur={[Function]}
          onFocus={[Function]}
          striped={true}
        >
          <tr
            className="c-table__row c-table__row--stripe "
            data-garden-id="tables.row"
            data-garden-version="version"
            onBlur={[Function]}
            onFocus={[Function]}
            tabIndex={-1}
          />
        </Row__StyledRow>
      </Row>
    </tbody>
  </table>
</ExampleWrapper>
`;
